[3.x] Fixed unnecessary bvh tree updates when calling set_pairable

pouleyKetchoupp opened the pull request:
Dynamic BVH doesn't update the tree anymore when calling set_pairable with no parameter change.

Also modified Godot Physics broadphase to create objects directly with pairable (static) set correctly to make use of this optimization for the BVH broadphase.

*Note:* Octree broadphase doesn't use this optimization because it forces an update on move, so passing the proper AABB and static parameters on creation would cause the tree to update twice.

---

Extracted from PR #48314 (see https://github.com/godotengine/godot/pull/48314#issuecomment-829898126)

This PR is 3.x only for now, because the dynamic BVH has been implemented only on this branch. Once everything is validated, it can be all ported to master.

pouleyKetchoupp committed:
Fixed unnecessary bvh tree updates when calling set_pairable

Dynamic BVH doesn't update the tree anymore when calling set_pairable
with no parameter change.

Also modified Godot Physics broadphase to create objects directly with
pairable (static) set correctly to make use of this optimization for the
BVH broadphase.

Note: Octree broadphase doesn't use this optimization because it forces
an update on move, so passing the proper AABB and static parameters on
creation would cause the tree to update twice.

lawnjelly commented on core/math/bvh.h:
```diff
@@ -282,15 +282,15 @@ class BVH_Manager {
 	// prefer calling this directly as type safe
 	void set_pairable(const BVHHandle &p_handle, bool p_pairable, uint32_t p_pairable_type, uint32_t p_pairable_mask) {
-		tree.item_set_pairable(p_handle, p_pairable, p_pairable_type, p_pairable_mask);
+		bool needs_update = tree.item_set_pairable(p_handle, p_pairable, p_pairable_type, p_pairable_mask);
```
It might be useful to put a comment above saying something like:
`// returns true if the pairing state has changed and the item is active`
Just to make it clear what is happening without needing to read the sub function.

lawnjelly commented:
With the `set_pairable` noop optimization, the only thing to watch for is if this is used as an initialization, and the collision check was assumed afterward. I've got a feeling there was an optimization during object creation to reduce to a single collision check somewhere which may be relevant,  I'll try and find it.

pouleyKetchoupp commented:
@lawnjelly Good catch! I've updated the code with option 1) as it does seem the most straightforward way to handle it. I've added a parameter `p_force_collision_check` that defaults to `true`, so there should be no change from the previous logic on the render side.

lawnjelly commented on core/math/bvh_public.inc:
```diff
@@ -328,6 +334,8 @@ void item_set_pairable(const BVHHandle &p_handle, bool p_pairable, uint32_t p_pa
 		// always keep this up to date
 		ex.pairable = p_pairable;
 	}
+
```
I'm not sure the logic in this function appears exactly the same depending on the quick reject (although it may produce the same result, it is confusing to follow).

Actually given how little the cost of the full logic and the potential for bugs like this (and future bugs), I think it is worth running the full logic in both cases, the only relevant thing is whether the function returns true or false (and thus runs the collision check which is the expensive bit). That is unless we have evidence that this is a bottleneck from profiling.

lawnjelly commented on core/math/bvh.h:
```diff
@@ -281,16 +281,17 @@ class BVH_Manager {
 
-			if (get_active(p_handle)) {
+			if (needs_update || (p_force_collision_check && tree.item_get_active(p_handle))) {
 				// when the pairable state changes, we need to force a collision check because newly pairable
```
This can use the `get_active(p_handle)` shortcut (if it is there we might as well use it for less verbosity, and encapsulation).

lawnjelly replied:
In fact I suspect for clarity it may be worth keeping the `get_active` check here in both cases, and having the bool returned be a `something_changed`. Unless profiling shows it to be the case, I don't think combining the two in one bool gains much, at the expensive of lowered clarity.

lawnjelly commented:
I've suggested a couple more tweaks.

Although the BVH should be fast, there appear to be some changes here which could make maintenance / debugging more difficult. Therefore to justify those it would be nice to see profiling info, for with and without those changes to see the benefit. If they give say 5% increase in speed then it is worth the sacrifice, but it could be that with compiler optimization they make < 0.1% difference, and so we need to balance the desire for optimization with clarity and maintenance.

The main thing it seems is avoiding the collision checks, and in practice I suspect everything else will be dwarfed by this.

pouleyKetchoupp commented:
@lawnjelly Thanks! I've rebased and made the changes based on your suggestions. I totally agree code quality should have priority unless profiling shows some optimizations are needed. I've tested again and I'm getting the same results as before, so the performance gain was negligible.

lawnjelly reviewed (approved):
Looks good! :+1:

akien-mga commented:
Thanks!
